File: arboles.py
from collections import deque

from nodos import NodoArbol


class ArbolBinario:
    def __init__(self):
        self.raiz = None

    # Inserta por niveles
    def insertar(self, valor):
        nuevo = NodoArbol(valor)
        if self.raiz is None:
            self.raiz = nuevo
            print("Nodo insertado.")
            return
        cola = deque([self.raiz])
        while cola:
            actual = cola.popleft()
            if actual.izquierda is None:
                actual.izquierda = nuevo
                print("Nodo insertado.")
                return
            cola.append(actual.izquierda)
            if actual.derecha is None:
                actual.derecha = nuevo
                print("Nodo insertado.")
                return
            cola.append(actual.derecha)

    # Busca por niveles
    def buscar(self, valor):
        if self.raiz is None:
            print("Arbol vacio.")
            return
        cola = deque([self.raiz])
        while cola:
            actual = cola.popleft()
            if actual.valor == valor:
                print("Encontrado.")
                return
            if actual.izquierda:
                cola.append(actual.izquierda)
            if actual.derecha:
                cola.append(actual.derecha)
        print("No encontrado.")

    # Recorre izquierda raiz derecha
    def inorder(self):
        resultado = []

        def recorrer(nodo):
            if nodo:
                recorrer(nodo.izquierda)
                resultado.append(nodo.valor)
                recorrer(nodo.derecha)

        recorrer(self.raiz)
        print(f"Inorder: {resultado}")

    # Recorre raiz izquierda derecha
    def preorder(self):
        resultado = []

        def recorrer(nodo):
            if nodo:
                resultado.append(nodo.valor)
                recorrer(nodo.izquierda)
                recorrer(nodo.derecha)

        recorrer(self.raiz)
        print(f"Preorder: {resultado}")

    # Recorre izquierda derecha raiz
    def postorder(self):
        resultado = []

        def recorrer(nodo):
            if nodo:
                recorrer(nodo.izquierda)
                recorrer(nodo.derecha)
                resultado.append(nodo.valor)

        recorrer(self.raiz)
        print(f"Postorder: {resultado}")


class ArbolBST:
    def __init__(self):
        self.raiz = None

    # Inserta respetando el orden
    def insertar(self, valor):
        self.raiz = self._insertar(self.raiz, valor)
        print("Nodo insertado.")

    # Inserta de forma recursiva
    def _insertar(self, nodo, valor):
        if nodo is None:
            return NodoArbol(valor)
        if valor < nodo.valor:
            nodo.izquierda = self._insertar(nodo.izquierda, valor)
        elif valor > nodo.valor:
            nodo.derecha = self._insertar(nodo.derecha, valor)
        return nodo

    # Busca usando el orden del BST
    def buscar(self, valor):
        actual = self.raiz
        while actual:
            if valor == actual.valor:
                print("Encontrado.")
                return
            if valor < actual.valor:
                actual = actual.izquierda
            else:
                actual = actual.derecha
        print("No encontrado.")

    # Recorre izquierda raiz derecha
    def inorder(self):
        resultado = []

        def recorrer(nodo):
            if nodo:
                recorrer(nodo.izquierda)
                resultado.append(nodo.valor)
                recorrer(nodo.derecha)

        recorrer(self.raiz)
        print(f"Inorder: {resultado}")

    # Recorre raiz izquierda derecha
    def preorder(self):
        resultado = []

        def recorrer(nodo):
            if nodo:
                resultado.append(nodo.valor)
                recorrer(nodo.izquierda)
                recorrer(nodo.derecha)

        recorrer(self.raiz)
        print(f"Preorder: {resultado}")

    # Recorre izquierda derecha raiz
    def postorder(self):
        resultado = []

        def recorrer(nodo):
            if nodo:
                recorrer(nodo.izquierda)
                recorrer(nodo.derecha)
                resultado.append(nodo.valor)

        recorrer(self.raiz)
        print(f"Postorder: {resultado}")

File: version_c/arboles.c
#include <stdio.h>
#include <stdlib.h>
#include "arboles.h"

// Crea un nodo
static NodoArbol *crear_nodo(int valor) {
    NodoArbol *nuevo = (NodoArbol *) malloc(sizeof(NodoArbol));
    nuevo->valor = valor;
    nuevo->izquierda = NULL;
    nuevo->derecha = NULL;
    return nuevo;
}

// Prepara el arbol
void inicializar_arbol_binario(ArbolBinario *arbol) {
    arbol->raiz = NULL;
}

// Inserta por niveles
void insertar_arbol_binario(ArbolBinario *arbol, int valor) {
    NodoArbol *nuevo = crear_nodo(valor);
    NodoArbol *cola[100];
    int frente = 0;
    int final = 0;

    if (arbol->raiz == NULL) {
        arbol->raiz = nuevo;
        printf("Nodo insertado\n");
        return;
    }

    cola[final++] = arbol->raiz;
    while (frente < final) {
        NodoArbol *actual = cola[frente++];
        if (actual->izquierda == NULL) {
            actual->izquierda = nuevo;
            printf("Nodo insertado\n");
            return;
        }
        cola[final++] = actual->izquierda;

        if (actual->derecha == NULL) {
            actual->derecha = nuevo;
            printf("Nodo insertado\n");
            return;
        }
        cola[final++] = actual->derecha;
    }
}

// Busca por niveles
void buscar_arbol_binario(ArbolBinario *arbol, int valor) {
    NodoArbol *cola[100];
    int frente = 0;
    int final = 0;

    if (arbol->raiz == NULL) {
        printf("Arbol vacio\n");
        return;
    }

    cola[final++] = arbol->raiz;
    while (frente < final) {
        NodoArbol *actual = cola[frente++];
        if (actual->valor == valor) {
            printf("Encontrado\n");
            return;
        }
        if (actual->izquierda != NULL) {
            cola[final++] = actual->izquierda;
        }
        if (actual->derecha != NULL) {
            cola[final++] = actual->derecha;
        }
    }
    printf("No encontrado\n");
}

// Recorre izquierda raiz derecha
void inorder_arbol(NodoArbol *raiz) {
    if (raiz != NULL) {
        inorder_arbol(raiz->izquierda);
        printf("%d ", raiz->valor);
        inorder_arbol(raiz->derecha);
    }
}

// Recorre raiz izquierda derecha
void preorder_arbol(NodoArbol *raiz) {
    if (raiz != NULL) {
        printf("%d ", raiz->valor);
        preorder_arbol(raiz->izquierda);
        preorder_arbol(raiz->derecha);
    }
}

// Recorre izquierda derecha raiz
void postorder_arbol(NodoArbol *raiz) {
    if (raiz != NULL) {
        postorder_arbol(raiz->izquierda);
        postorder_arbol(raiz->derecha);
        printf("%d ", raiz->valor);
    }
}

// Libera memoria
void liberar_arbol(NodoArbol *raiz) {
    if (raiz != NULL) {
        liberar_arbol(raiz->izquierda);
        liberar_arbol(raiz->derecha);
        free(raiz);
    }
}

// Prepara el arbol
void inicializar_bst(ArbolBST *arbol) {
    arbol->raiz = NULL;
}

// Inserta en BST
static NodoArbol *insertar_bst_rec(NodoArbol *raiz, int valor) {
    if (raiz == NULL) {
        return crear_nodo(valor);
    }
    if (valor < raiz->valor) {
        raiz->izquierda = insertar_bst_rec(raiz->izquierda, valor);
    } else if (valor > raiz->valor) {
        raiz->derecha = insertar_bst_rec(raiz->derecha, valor);
    }
    return raiz;
}

// Inserta en BST
void insertar_bst(ArbolBST *arbol, int valor) {
    arbol->raiz = insertar_bst_rec(arbol->raiz, valor);
    printf("Nodo insertado\n");
}

// Busca en BST
void buscar_bst(ArbolBST *arbol, int valor) {
    NodoArbol *actual = arbol->raiz;
    while (actual != NULL) {
        if (valor == actual->valor) {
            printf("Encontrado\n");
            return;
        }
        if (valor < actual->valor) {
            actual = actual->izquierda;
        } else {
            actual = actual->derecha;
        }
    }
    printf("No encontrado\n");
}
